### ISSUE TITLE
pandas boxplot shows a strange output if by=df['datetime_column'].dt.functions

### ISSUE THREAD

**Issue by sawoha** [pandas-dev/pandas#23007]
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd

df['value'] = pd.DataFrame(np.random.random((200,1)))
df['date'] = pd.date_range('2000-1-1', periods=200, freq='D')

df.boxplot(column='value',by=df['date'].dt.month)
```
#### Problem description
The results shown here explains a lot so I am not going to elaborate much.
As far as I tested, this only happens if datetime function is used for grouping values to draw multiple boxplots in one plot.
![untitled](https://user-images.githubusercontent.com/43888085/46546309-42ed6200-c8fb-11e8-9602-84e195818635.png)

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.2.0
Cython: None
numpy: 1.15.2
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: 1.8.1
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

**Comment by mroeschke** [comment]
So you're passing an array-like value to `by`, and it seems like the title is using the `by` value as part of the title and xlabel hence why you are seeing the array repr cloud the plot.

I think it's reasonable to suppress array inputs from being included in the supporting plot text, but what would be a suitable alternative label (maybe we just don't include a label)?

cc @TomAugspurger 

FYI, to get a better looking plot, create a new `'month'` column and then specify `by='month'` in the boxplot

```
df['month'] = df['date'].dt.month
df.boxplot(column='value', by='month')
```

**Comment by sawoha** [comment]
> FYI, to get a better looking plot, create a new `'month'` column and then specify `by='month'` in the boxplot
> 
> ```
> df['month'] = df['date'].dt.month
> df.boxplot(column='value', by='month')
> ```

Yes, I am currently circumventing this issue by creating a new column dedicated to month/year/day/etc.

However, I am still confused about why dataframe.hist() and dataframe.plot.hist() show different behaviors. 

For example, the following code works fine without any issue.

df.hist(column='value',by=df['date'].dt.month)
![untitled](https://user-images.githubusercontent.com/43888085/46565768-f0856300-c945-11e8-855a-1d9d91934c68.png)

However, if I use dataframe.plot.hist to the same job, it returns non-grouped result
= identical to the results of df['value'].plot.hist() or df.hist(column='value')

df['value'].plot.hist(by=df['date'].dt.month)
![index](https://user-images.githubusercontent.com/43888085/46565902-1a3f8980-c948-11e8-9b41-f4048d66d64c.png)

**Comment by mroeschke** [comment]
The docstring for `DataFrame.hist` states:


> A histogram is a representation of the distribution of data. This function calls matplotlib.pyplot.hist(), on each series in the DataFrame, resulting in one histogram per column

So in that case you'll get one plot per column _per group_

As opposed to `df['value']` which is a series, so it will just produce one plot since the base data is 1D

**Comment by sawoha** [comment]
@mroeschke , thanks again for the clarification.

 I am still wondering about why the result of `df['value'].hist(by=df['date'].dt.month)` is different from `df['value'].plot.hist(by=df['date'].dt.month)`, but this is not relevant to the topic anymore.

**Comment by TomAugspurger** [comment]
I think `by=array` should essentially behave the same as assigning that array to a column and taking passing the column name to `by`.
